commit ffb24b3675d460afa58702205518195681b3422c
Author: Quentin Smith <quentin@mit.edu>
Date:   Thu Apr 3 01:27:28 2025 -0400

    Support config.yaml SSO apps

diff --git a/inventree-src/src/backend/InvenTree/InvenTree/auth_overrides.py b/inventree-src/src/backend/InvenTree/InvenTree/auth_overrides.py
index 9b2387983..dea294e74 100644
--- a/inventree-src/src/backend/InvenTree/InvenTree/auth_overrides.py
+++ b/inventree-src/src/backend/InvenTree/InvenTree/auth_overrides.py
@@ -91,7 +91,7 @@ def registration_enabled():
         get_global_setting('LOGIN_ENABLE_REG')
         or InvenTree.sso.sso_registration_enabled()
     ):
-        if is_email_configured():
+        if is_email_configured() or not get_global_setting('LOGIN_ENABLE_PWD_FORGOT'):
             return True
         else:
             logger.warning(
diff --git a/inventree-src/src/backend/InvenTree/InvenTree/sso.py b/inventree-src/src/backend/InvenTree/InvenTree/sso.py
index c1842d70d..6c74afc6a 100644
--- a/inventree-src/src/backend/InvenTree/InvenTree/sso.py
+++ b/inventree-src/src/backend/InvenTree/InvenTree/sso.py
@@ -18,20 +18,19 @@ logger = structlog.get_logger('inventree')
 def get_provider_app(provider):
     """Return the SocialApp object for the given provider."""
     from allauth.socialaccount.models import SocialApp
+    from allauth.socialaccount.adapter import get_adapter
 
+    adapter = get_adapter()
     try:
-        apps = SocialApp.objects.filter(provider__iexact=provider.id)
+        apps = adapter.list_apps(None)
+        logger.info("SSO SocialApps: %r", apps)
+        return adapter.get_app(None, provider.id)
     except SocialApp.DoesNotExist:
         logger.warning("SSO SocialApp not found for provider '%s'", provider.id)
         return None
-
-    if apps.count() > 1:
+    except SocialApp.MultipleObjectsReturned:
         logger.warning("Multiple SocialApps found for provider '%s'", provider.id)
-
-    if apps.count() == 0:
-        logger.warning("SSO SocialApp not found for provider '%s'", provider.id)
-
-    return apps.first()
+        return None
 
 
 def check_provider(provider):
