diff --git a/board/raspberrypi/rpi/rpi.env b/board/raspberrypi/rpi/rpi.env
index 9ac9d6768ca..c0225358345 100644
--- a/board/raspberrypi/rpi/rpi.env
+++ b/board/raspberrypi/rpi/rpi.env
@@ -5,9 +5,9 @@
 dhcpuboot=usb start; dhcp u-boot.uimg; bootm
 
 /* Environment */
-stdin=serial,usbkbd
-stdout=serial,vidconsole
-stderr=serial,vidconsole
+stdin=nulldev,usbkbd
+stdout=nulldev,vidconsole
+stderr=nulldev,vidconsole
 
 /* DFU over USB/UDC */
 #ifdef CONFIG_CMD_DFU
