commit f1510311a27418982581858ce568d20142f03642
Author: Quentin Smith <quentin@mit.edu>
Date:   Fri Apr 28 02:55:03 2023 -0400

    Use esphome object ID as default entity ID

diff --git a/homeassistant/components/esphome/__init__.py b/homeassistant/components/esphome/__init__.py
index 4658893d37..93cb761d4c 100644
--- a/homeassistant/components/esphome/__init__.py
+++ b/homeassistant/components/esphome/__init__.py
@@ -56,6 +56,7 @@ from homeassistant.helpers.issue_registry import (
 )
 from homeassistant.helpers.service import async_set_service_schema
 from homeassistant.helpers.template import Template
+from homeassistant.util import slugify
 
 from .bluetooth import async_connect_scanner
 from .const import DOMAIN
@@ -743,6 +744,7 @@ class EsphomeEntity(Entity, Generic[_InfoT, _StateT]):
     """Define a base esphome entity."""
 
     _attr_should_poll = False
+    _entity_id = None
 
     def __init__(
         self,
@@ -759,6 +761,19 @@ class EsphomeEntity(Entity, Generic[_InfoT, _StateT]):
         if entry_data.device_info is not None and entry_data.device_info.friendly_name:
             self._attr_has_entity_name = True
 
+    @property
+    def entity_id(self):
+        # Can't set this in __init__ because self._static_info doesn't exist yet.
+        if self._entity_id:
+            return self._entity_id
+        if self._static_info.object_id:
+            return "%s.%s" % (self._component_key, slugify(self._static_info.object_id))
+        return None
+
+    @entity_id.setter
+    def entity_id(self, entity_id):
+        self._entity_id = entity_id
+
     async def async_added_to_hass(self) -> None:
         """Register callbacks."""
         self.async_on_remove(
