diff --git a/homeassistant/components/esphome/entity.py b/homeassistant/components/esphome/entity.py
index 7b02680afee..62a72f1ffe1 100644
--- a/homeassistant/components/esphome/entity.py
+++ b/homeassistant/components/esphome/entity.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 from collections.abc import Awaitable, Callable, Coroutine
 import functools
 import math
-from typing import TYPE_CHECKING, Any, Concatenate, Generic, TypeVar, cast
+from typing import TYPE_CHECKING, Any, Concatenate, Generic, Optional, ParamSpec, TypeVar, cast
 
 from aioesphomeapi import (
     APIConnectionError,
@@ -28,11 +28,12 @@ from homeassistant.helpers import (
 from homeassistant.helpers.device_registry import DeviceInfo
 from homeassistant.helpers.entity import Entity
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import slugify
 
 from .const import DOMAIN
 
 # Import config flow so that it's added to the registry
-from .entry_data import ESPHomeConfigEntry, RuntimeEntryData
+from .entry_data import ESPHomeConfigEntry, RuntimeEntryData, INFO_TYPE_TO_PLATFORM
 from .enum_mapper import EsphomeEnumMapper
 
 _InfoT = TypeVar("_InfoT", bound=EntityInfo)
@@ -210,6 +211,7 @@ class EsphomeEntity(EsphomeBaseEntity, Generic[_InfoT, _StateT]):
     _static_info: _InfoT
     _state: _StateT
     _has_state: bool
+    _entity_id: Optional[str] = None
 
     def __init__(
         self,
@@ -242,6 +244,20 @@ class EsphomeEntity(EsphomeBaseEntity, Generic[_InfoT, _StateT]):
             # an empty string so we drop the object_id.
             self.entity_id = f"{domain}.{device_info.name}"
 
+    @property
+    def entity_id(self):
+        # Can't set this in __init__ because self._static_info doesn't exist yet.
+        if self._entity_id:
+            return self._entity_id
+        if self._static_info.object_id:
+            platform = INFO_TYPE_TO_PLATFORM[type(self._static_info)]
+            return "%s.%s" % (platform, slugify(self._static_info.object_id))
+        return None
+    @entity_id.setter
+    def entity_id(self, entity_id):
+        self._entity_id = entity_id
+
+
     async def async_added_to_hass(self) -> None:
         """Register callbacks."""
         entry_data = self._entry_data
