commit 8c982b9341c9c39183b2eec15e0902f7897f406f
Author: Quentin Smith <quentin@mit.edu>
Date:   Fri Apr 28 02:55:03 2023 -0400

    Use esphome object ID as default entity ID

diff --git a/homeassistant/components/esphome/__init__.py b/homeassistant/components/esphome/__init__.py
index 4658893d37..a14ee33e8c 100644
--- a/homeassistant/components/esphome/__init__.py
+++ b/homeassistant/components/esphome/__init__.py
@@ -56,6 +56,7 @@ from homeassistant.helpers.issue_registry import (
 )
 from homeassistant.helpers.service import async_set_service_schema
 from homeassistant.helpers.template import Template
+from homeassistant.util import slugify
 
 from .bluetooth import async_connect_scanner
 from .const import DOMAIN
@@ -758,6 +759,8 @@ class EsphomeEntity(Entity, Generic[_InfoT, _StateT]):
         self._state_type = state_type
         if entry_data.device_info is not None and entry_data.device_info.friendly_name:
             self._attr_has_entity_name = True
+        if self._static_info.object_id:
+            self.entity_id = "%s.%s" % (component_key, slugify(self._static_info.object_id))
 
     async def async_added_to_hass(self) -> None:
         """Register callbacks."""
