diff --git a/syncoid b/syncoid
index 680afbe..dcfc40a 100755
--- a/syncoid
+++ b/syncoid
@@ -51,15 +51,17 @@ if (length $args{'sendoptions'}) {
 		exit 127;
 	}
 
-	if (defined $args{'recursive'}) {
-		foreach my $option(@sendoptions) {
-			if ($option->{option} eq 'R') {
-				writelog('WARN', "invalid argument combination, zfs send -R and --recursive aren't compatible!");
-				pod2usage(2);
-				exit 127;
-			}
+	foreach my $option(@sendoptions) {
+		if ($option->{option} eq 'R') {
+			$args{'native-recursion'} = 1;
 		}
 	}
+
+	if (defined $args{'recursive'} && defined $args{'native-recursion'}) {
+		writelog('WARN', "invalid argument combination, zfs send -R and --recursive aren't compatible!");
+		pod2usage(2);
+		exit 127;
+	}
 }
 
 my @recvoptions = ();
@@ -342,7 +344,6 @@ sub syncdataset {
 	my $stdout;
 	my $exit;
 
-	my $sourcefsescaped = escapeshellparam($sourcefs);
 	my $targetfsescaped = escapeshellparam($targetfs);
 
 	# if no rollbacks are allowed, disable forced receive
@@ -411,6 +412,12 @@ sub syncdataset {
 	my $newsyncsnap;
 	my $matchingsnap;
 
+	my $hostid = hostname();
+	my $holdname = "syncoid\_$identifier$hostid";
+	if (defined $args{'native-recursion'}) {
+		$holdname = "-r $holdname";
+	}
+
 	# skip snapshot checking/creation in case of resumed receive
 	if (!defined($receivetoken)) {
 		# build hashes of the snaps on the source and target filesystems.
@@ -455,7 +462,12 @@ sub syncdataset {
 			}
 		}
 	}
-	my $newsyncsnapescaped = escapeshellparam($newsyncsnap);
+
+	if ($newsyncsnap && defined $args{'use-hold'}) {
+		# Hold the new snapshot before we start syncing, to ensure it's not deleted
+		# before we have a chance to hold it.
+		runholdcmd($sourcehost, $sourcesudocmd, $sourcefs, $newsyncsnap, "hold", $holdname);
+	}
 
 	# there is currently (2014-09-01) a bug in ZFS on Linux
 	# that causes readonly to always show on if it's EVER
@@ -665,8 +677,6 @@ sub syncdataset {
 			writelog('INFO', "no snapshots on source newer than $newsyncsnap on target. Nothing to do, not syncing.");
 			return 0;
 		} else {
-			my $matchingsnapescaped = escapeshellparam($matchingsnap);
-
 			my $nextsnapshot = 0;
 
 			if ($bookmark) {
@@ -703,7 +713,6 @@ sub syncdataset {
 					};
 
 					$matchingsnap = $nextsnapshot;
-					$matchingsnapescaped = escapeshellparam($matchingsnap);
 				} else {
 					($exit, $stdout) = syncbookmark($sourcehost, $sourcefs, $targethost, $targetfs, $bookmark, $newsyncsnap);
 
@@ -792,37 +801,14 @@ sub syncdataset {
 	# if "--use-hold" parameter is used set hold on newsync snapshot and remove hold on matching snapshot both on source and target
 	# hold name: "syncoid" + identifier + hostname -> in case of replication to multiple targets separate holds can be set for each target by assinging different identifiers to each target. Only if all targets have been replicated all syncoid holds are removed from the matching snapshot and it can be removed
 	if (defined $args{'use-hold'}) {
-		my $holdcmd;
-		my $holdreleasecmd;
-		my $hostid = hostname();
-		my $matchingsnapescaped = escapeshellparam($matchingsnap);
-		my $holdname = "syncoid\_$identifier$hostid";
-		if ($sourcehost ne '') {
-			$holdcmd = "$sshcmd $sourcehost " . escapeshellparam("$sourcesudocmd $zfscmd hold $holdname $sourcefsescaped\@$newsyncsnapescaped");
-			$holdreleasecmd = "$sshcmd $sourcehost " . escapeshellparam("$sourcesudocmd $zfscmd release $holdname $sourcefsescaped\@$matchingsnapescaped");
-		} else {
-			$holdcmd = "$sourcesudocmd $zfscmd hold $holdname $sourcefsescaped\@$newsyncsnapescaped";
-			$holdreleasecmd = "$sourcesudocmd $zfscmd release $holdname $sourcefsescaped\@$matchingsnapescaped";
-		}
-		writelog('DEBUG', "Set new hold on source: $holdcmd");
-		system($holdcmd) == 0 or warn "WARNING: $holdcmd failed: $?";
 		# Do hold release only if matchingsnap exists
 		if ($matchingsnap) {
-			writelog('DEBUG', "Release old hold on source: $holdreleasecmd");
-			system($holdreleasecmd) == 0 or warn "WARNING: $holdreleasecmd failed: $?";
+			runholdcmd($sourcehost, $sourcesudocmd, $sourcefs, $matchingsnap, "release", $holdname);
 		}
-		if ($targethost ne '') {
-			$holdcmd = "$sshcmd $targethost " . escapeshellparam("$targetsudocmd $zfscmd hold $holdname $targetfsescaped\@$newsyncsnapescaped");
-			$holdreleasecmd = "$sshcmd $targethost " . escapeshellparam("$targetsudocmd $zfscmd release $holdname $targetfsescaped\@$matchingsnapescaped");
-		} else {
-			$holdcmd = "$targetsudocmd $zfscmd hold $holdname $targetfsescaped\@$newsyncsnapescaped";				$holdreleasecmd = "$targetsudocmd $zfscmd release $holdname $targetfsescaped\@$matchingsnapescaped";
-		}
-		writelog('DEBUG', "Set new hold on target: $holdcmd");
-		system($holdcmd) == 0 or warn "WARNING: $holdcmd failed: $?";
+		runholdcmd($targethost, $targetsudocmd, $targetfs, $newsyncsnap, "hold", $holdname);
 		# Do hold release only if matchingsnap exists
 		if ($matchingsnap) {
-			writelog('DEBUG', "Release old hold on target: $holdreleasecmd");
-			system($holdreleasecmd) == 0 or warn "WARNING: $holdreleasecmd failed: $?";
+			runholdcmd($targethost, $targetsudocmd, $targetfs, $matchingsnap, "release", $holdname);
 		}
 	}
 	if (defined $args{'no-sync-snap'}) {
@@ -885,6 +871,21 @@ sub syncdataset {
 
 } # end syncdataset()
 
+sub runholdcmd {
+	my ($host, $sudocmd, $fs, $snap, $action, $holdname) = @_;
+
+	my $fsescaped = escapeshellparam($fs);
+	my $snapescaped = escapeshellparam($snap);
+
+	my $holdcmd = "$sudocmd $zfscmd $action $holdname $fsescaped\@$snapescaped";
+
+	if ($host ne '') {
+		$holdcmd = "$sshcmd $host " . escapeshellparam($holdcmd);
+	}
+	writelog('DEBUG', "Modify hold: $holdcmd");
+	system($holdcmd) == 0 or warn "WARNING: $holdcmd failed: $?";
+}
+
 # Return codes:
 #  0 - ZFS send/receive completed without errors
 #  1 - ZFS target is currently in receive
@@ -1749,7 +1750,11 @@ sub newsyncsnap {
 	my $hostid = hostname();
 	my %date = getdate();
 	my $snapname = "syncoid\_$identifier$hostid\_$date{'stamp'}";
-	my $snapcmd = "$rhost $mysudocmd $zfscmd snapshot $fsescaped\@$snapname\n";
+	my $snapcmd = "$rhost $mysudocmd $zfscmd snapshot";
+	if (defined $args{'native-recursion'}) {
+		$snapcmd .= " -r";
+	}
+	$snapcmd .= " $fsescaped\@$snapname\n";
 	writelog('DEBUG', "creating sync snapshot using \"$snapcmd\"...");
 	system($snapcmd) == 0 or do {
 		writelog('CRITICAL', "$snapcmd failed: $?");
