diff --git a/src/conf/domain_conf.c b/src/conf/domain_conf.c
index 45235d74bc..b283906715 100644
--- a/src/conf/domain_conf.c
+++ b/src/conf/domain_conf.c
@@ -902,6 +902,7 @@ VIR_ENUM_IMPL(virDomainVideo,
               "none",
               "bochs",
               "ramfb",
+              "nubus-macfb",
 );
 
 VIR_ENUM_IMPL(virDomainVideoVGAConf,
@@ -13720,6 +13721,7 @@ virDomainVideoDefaultRAM(const virDomainDef *def,
     case VIR_DOMAIN_VIDEO_TYPE_GOP:
     case VIR_DOMAIN_VIDEO_TYPE_NONE:
     case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+    case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
     case VIR_DOMAIN_VIDEO_TYPE_LAST:
     default:
         return 0;
diff --git a/src/conf/domain_conf.h b/src/conf/domain_conf.h
index 91f57de0f1..13d5d84848 100644
--- a/src/conf/domain_conf.h
+++ b/src/conf/domain_conf.h
@@ -1870,6 +1870,7 @@ typedef enum {
     VIR_DOMAIN_VIDEO_TYPE_NONE,
     VIR_DOMAIN_VIDEO_TYPE_BOCHS,
     VIR_DOMAIN_VIDEO_TYPE_RAMFB,
+    VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB,
 
     VIR_DOMAIN_VIDEO_TYPE_LAST
 } virDomainVideoType;
diff --git a/src/libxl/libxl_conf.c b/src/libxl/libxl_conf.c
index 3bb5c1e591..a5a82b506b 100644
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -2471,6 +2471,7 @@ libxlMakeVideo(virDomainDef *def, libxl_domain_config *d_config)
         case VIR_DOMAIN_VIDEO_TYPE_NONE:
         case VIR_DOMAIN_VIDEO_TYPE_BOCHS:
         case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+        case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
         case VIR_DOMAIN_VIDEO_TYPE_LAST:
         default:
             virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
diff --git a/src/libxl/libxl_domain.c b/src/libxl/libxl_domain.c
index 51dc3edcc2..fcb2e825e2 100644
--- a/src/libxl/libxl_domain.c
+++ b/src/libxl/libxl_domain.c
@@ -216,6 +216,7 @@ libxlDomainDeviceDefPostParse(virDomainDeviceDef *dev,
                 case VIR_DOMAIN_VIDEO_TYPE_NONE:
                 case VIR_DOMAIN_VIDEO_TYPE_BOCHS:
                 case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+                case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
                 case VIR_DOMAIN_VIDEO_TYPE_LAST:
                     break;
             }
diff --git a/src/qemu/qemu_capabilities.c b/src/qemu/qemu_capabilities.c
index 61e44b398a..20074f05d8 100644
--- a/src/qemu/qemu_capabilities.c
+++ b/src/qemu/qemu_capabilities.c
@@ -768,6 +768,7 @@ VIR_ENUM_IMPL(virQEMUCaps,
               /* 495 */
               "blockdev-mirror.target-is-zero", /* QEMU_CAPS_BLOCKDEV_MIRROR_TARGET_IS_ZERO */
               "object-monitor-qmp", /* QEMU_CAPS_OBJECT_MONITOR_QMP */
+              "nubus-macfb", /* QEMU_CAPS_DEVICE_NUBUS_MACFB */
     );
 
 
@@ -1481,6 +1482,7 @@ struct virQEMUCapsStringFlags virQEMUCapsObjectTypes[] = {
     { "uefi-vars-x64", QEMU_CAPS_DEVICE_UEFI_VARS },
     { "uefi-vars-sysbus", QEMU_CAPS_DEVICE_UEFI_VARS },
     { "monitor-qmp", QEMU_CAPS_OBJECT_MONITOR_QMP },
+    { "nubus-macfb", QEMU_CAPS_DEVICE_NUBUS_MACFB },
 };
 
 
@@ -6924,6 +6926,8 @@ virQEMUCapsFillDomainDeviceVideoCaps(virQEMUCaps *qemuCaps,
         VIR_DOMAIN_CAPS_ENUM_SET(dev->modelType, VIR_DOMAIN_VIDEO_TYPE_BOCHS);
     if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_DEVICE_RAMFB))
         VIR_DOMAIN_CAPS_ENUM_SET(dev->modelType, VIR_DOMAIN_VIDEO_TYPE_RAMFB);
+    if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_DEVICE_NUBUS_MACFB))
+        VIR_DOMAIN_CAPS_ENUM_SET(dev->modelType, VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB);
 }
 
 
diff --git a/src/qemu/qemu_capabilities.h b/src/qemu/qemu_capabilities.h
index c846771d50..45f81626ee 100644
--- a/src/qemu/qemu_capabilities.h
+++ b/src/qemu/qemu_capabilities.h
@@ -742,6 +742,7 @@ typedef enum { /* virQEMUCapsFlags grouping marker for syntax-check */
     /* 495 */
     QEMU_CAPS_BLOCKDEV_MIRROR_TARGET_IS_ZERO, /* 'blockdev-mirror' supports 'target-is-zero' */
     QEMU_CAPS_OBJECT_MONITOR_QMP, /* 'monitor-qmp' -object */
+    QEMU_CAPS_DEVICE_NUBUS_MACFB, /* -device nubus-macfb */
 
     QEMU_CAPS_LAST /* this must always be the last item */
 } virQEMUCapsFlags;
diff --git a/src/qemu/qemu_command.c b/src/qemu/qemu_command.c
index a76a034158..5e02d37930 100644
--- a/src/qemu/qemu_command.c
+++ b/src/qemu/qemu_command.c
@@ -89,6 +89,7 @@ VIR_ENUM_IMPL(qemuVideo,
               "" /* 'none' doesn't make sense here */,
               "bochs-display",
               "", /* ramfb can't be used with -vga */
+              "", /* nubus-macfb can't be used with -vga */
 );
 
 VIR_ENUM_IMPL(qemuSoundCodec,
@@ -762,6 +763,9 @@ qemuDeviceVideoGetModel(const virDomainVideoDef *video,
     case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
         return "ramfb";
 
+    case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
+        return "nubus-macfb";
+
     case VIR_DOMAIN_VIDEO_TYPE_VIRTIO: {
         struct qemuDeviceVideoModelVirtioOpts *opts = &qemuDeviceVideoGetModelVirtioOptsTable[video->virtiodevice];
 
diff --git a/src/qemu/qemu_domain.c b/src/qemu/qemu_domain.c
index a4e5f92840..473758b241 100644
--- a/src/qemu/qemu_domain.c
+++ b/src/qemu/qemu_domain.c
@@ -7882,6 +7882,9 @@ qemuDomainSupportsPCI(const virDomainDef *def)
     if (def->os.arch == VIR_ARCH_SPARC)
         return false;
 
+    if (def->os.arch == VIR_ARCH_M68K)
+        return false;
+
     /* On all other architectures, PCI support is assumed to
      * be present */
     return true;
@@ -8850,6 +8853,10 @@ qemuDomainSupportsNicdev(virDomainDef *def,
         net->info.type != VIR_DOMAIN_DEVICE_ADDRESS_TYPE_PCI)
         return false;
 
+    /* m68k nics require legacy -net nic */
+    if (def->os.arch == VIR_ARCH_M68K)
+        return false;
+
     return true;
 }
 
diff --git a/src/qemu/qemu_domain_address.c b/src/qemu/qemu_domain_address.c
index a8292a9782..30f8c0cb85 100644
--- a/src/qemu/qemu_domain_address.c
+++ b/src/qemu/qemu_domain_address.c
@@ -931,6 +931,7 @@ qemuDomainDeviceCalculatePCIConnectFlags(virDomainDeviceDef *dev,
         case VIR_DOMAIN_VIDEO_TYPE_GOP:
         case VIR_DOMAIN_VIDEO_TYPE_NONE:
         case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+        case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
         case VIR_DOMAIN_VIDEO_TYPE_LAST:
             return 0;
         }
@@ -2346,7 +2347,8 @@ qemuDomainAssignDevicePCISlots(virDomainDef *def,
     /* Video devices */
     for (i = 0; i < def->nvideos; i++) {
         if (def->videos[i]->type == VIR_DOMAIN_VIDEO_TYPE_NONE ||
-            def->videos[i]->type == VIR_DOMAIN_VIDEO_TYPE_RAMFB)
+            def->videos[i]->type == VIR_DOMAIN_VIDEO_TYPE_RAMFB ||
+            def->videos[i]->type == VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB)
             continue;
 
         if (!virDeviceInfoPCIAddressIsWanted(&def->videos[i]->info))
diff --git a/src/qemu/qemu_monitor_json.c b/src/qemu/qemu_monitor_json.c
index af9a77ed3b..3cbef48b5c 100644
--- a/src/qemu/qemu_monitor_json.c
+++ b/src/qemu/qemu_monitor_json.c
@@ -1939,6 +1939,7 @@ qemuMonitorJSONUpdateVideoMemorySize(qemuMonitor *mon,
     case VIR_DOMAIN_VIDEO_TYPE_NONE:
     case VIR_DOMAIN_VIDEO_TYPE_BOCHS:
     case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+    case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
     case VIR_DOMAIN_VIDEO_TYPE_LAST:
         break;
     }
@@ -1988,6 +1989,7 @@ qemuMonitorJSONUpdateVideoVram64Size(qemuMonitor *mon,
     case VIR_DOMAIN_VIDEO_TYPE_NONE:
     case VIR_DOMAIN_VIDEO_TYPE_BOCHS:
     case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+    case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
     case VIR_DOMAIN_VIDEO_TYPE_LAST:
         break;
     }
diff --git a/src/qemu/qemu_postparse.c b/src/qemu/qemu_postparse.c
index 9eda2f6b99..8715ccec69 100644
--- a/src/qemu/qemu_postparse.c
+++ b/src/qemu/qemu_postparse.c
@@ -301,6 +301,8 @@ qemuDomainDefaultVideoDevice(const virDomainDef *def,
         return VIR_DOMAIN_VIDEO_TYPE_CIRRUS;
     if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_DEVICE_VGA))
         return VIR_DOMAIN_VIDEO_TYPE_VGA;
+    if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_DEVICE_NUBUS_MACFB))
+        return VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB;
     return VIR_DOMAIN_VIDEO_TYPE_DEFAULT;
 }
 
diff --git a/src/qemu/qemu_process.c b/src/qemu/qemu_process.c
index b2506edce0..29f5d3dacc 100644
--- a/src/qemu/qemu_process.c
+++ b/src/qemu/qemu_process.c
@@ -3298,6 +3298,7 @@ qemuProcessUpdateVideoRamSize(virDomainObj *vm,
         case VIR_DOMAIN_VIDEO_TYPE_NONE:
         case VIR_DOMAIN_VIDEO_TYPE_BOCHS:
         case VIR_DOMAIN_VIDEO_TYPE_RAMFB:
+        case VIR_DOMAIN_VIDEO_TYPE_NUBUS_MACFB:
         case VIR_DOMAIN_VIDEO_TYPE_LAST:
             break;
         }
