--- a/drivers/acpi/thermal.c	2024-03-15 14:19:29.000000000 -0400
+++ a/drivers/acpi/thermal.c	2024-03-30 19:35:55.075089786 -0400
@@ -858,11 +858,8 @@
 	if (hot_temp != THERMAL_TEMP_INVALID)
 		trip_count++;
 
-	if (!trip_count) {
+	if (!trip_count)
 		pr_warn(FW_BUG "No valid trip points!\n");
-		result = -ENODEV;
-		goto free_memory;
-	}
 
 	/* Get temperature [_TMP] (required). */
 	result = acpi_thermal_get_temperature(tz);
